store/cache/cache.go

```diff
@@ -0,0 +1,112 @@
+package cache
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/store/cachekv"
+	"github.com/cosmos/cosmos-sdk/store/types"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+var (
+	_ types.CommitKVStore             = (*CommitKVStoreCache)(nil)
+	_ types.MultiStorePersistentCache = (*CommitKVStoreCacheManager)(nil)
+
+	// DefaultCommitKVStoreCacheSize defines the persistent ARC cache size for a
+	// CommitKVStoreCache.
+	DefaultCommitKVStoreCacheSize uint = 1000
+)
+
+type (
+	// CommitKVStoreCache implements an inter-block (persistent) cache that wraps a
+	// CommitKVStore. Reads first hit the internal ARC (Adaptive Replacement Cache).
+	// During a cache miss, the read is delegated to the underlying CommitKVStore
+	// and cached. Deletes and writes always happen to both the cache and the
+	// CommitKVStore in a write-through manner. Caching performed in the
+	// CommitKVStore and below is completely irrelevant to this layer.
+	CommitKVStoreCache struct {
+		types.CommitKVStore
+		cache *lru.ARCCache
+	}
+
+	// CommitKVStoreCacheManager maintains a mapping from a StoreKey to a
+	// CommitKVStoreCache. Each CommitKVStore, per StoreKey, is meant to be used
+	// in an inter-block (persistent) manner and typically provided by a
+	// CommitMultiStore.
+	CommitKVStoreCacheManager struct {
+		cacheSize uint
```

Comment:
Would it be useful to be able to define cacheSize on a per-store basis?

For example, may not want a cache at all for accountStore. May want a bigger cache for govStore. A small but performant cache for params, etc.

Reply:
Yeah so in theory that may prove to be valuable, but I hesitate to add such functionality in this initial PR for the following reasons:

- Complicates the API (although frivolous) 
- Requires fine benchmarking to know the values for each desired respective store

I recommend we create these APIs in a subsequent PR in a non-breaking fashion.